bevy_lint/src/lints/suspicious/unit_in_bundle.rs

```diff
@@ -0,0 +1,415 @@
+///
+/// Each returned [`Ty`] is guaranteed to be a [`ty::TyKind::Param`].
+///
+/// # Example
+///
+/// If run on the following function, this function would return `A` and `C` because they both
+/// implement `Bundle`.
+///
+/// ```
+/// fn my_function<A: Bundle, B: Clone, C: Bundle + Clone>(_: A, _: B, _: C) {
+///     // ...
+/// }
+/// ```
+fn bundle_bounded_generics<'tcx>(cx: &LateContext<'tcx>, fn_id: DefId) -> Vec<Ty<'tcx>> {
+    let mut bundle_bounded_generics = Vec::new();
+
+    // Fetch the parameter environment for the function, which contains all generic trait bounds.
+    // (Such as the `T: Bundle` that we're looking for!) See
+    // <https://rustc-dev-guide.rust-lang.org/typing_parameter_envs.html> for more information.
+    let param_env = cx.tcx.param_env(fn_id);
+
+    for clause in param_env.caller_bounds() {
+        // We only want trait predicates, filtering out lifetimes and constants.
+        if let Some(trait_predicate) = clause.as_trait_clause()
+            // The `Bundle` trait doesn't require any bound vars, so we dispel the binder.
+            && let Some(trait_predicate) = trait_predicate.no_bound_vars()
+            && let ty::TraitPredicate {
+                trait_ref,
+                // Negative trait bounds, which are unstable, allow matching all types _except_
+                // those with a specific trait. We don't want that, however, so we only match
+                // positive trait bounds.
+                polarity: PredicatePolarity::Positive,
+            } = trait_predicate
+            // Only match `T: Bundle` predicates.
+            && paths::BUNDLE.matches(cx, trait_ref.def_id)
+        {
+            let self_ty = trait_ref.self_ty();
+
+            debug_assert!(
+                matches!(self_ty.kind(), ty::TyKind::Param(_)),
+                "type from trait bound was expected to be a type parameter",
+            );
+
+            // At this point, we've confirmed the predicate is `T: Bundle`! Add it to the list to
+            // be returned. :)
+            bundle_bounded_generics.push(trait_ref.self_ty());
+        }
+    }
+
+    bundle_bounded_generics
+}
+
+/// Represents the path to an item within a nested tuple.
+///
+/// # Example
+///
+/// Each number within the [`TuplePath`] represents an index into the tuple. An empty path
+/// represents the root tuple, while a path of `TuplePath([0])` represents the first item within
+/// that tuple.
+///
+/// ```ignore
+/// // TuplePath([])
+/// (
+///     // TuplePath([0])
+///     Name::new("Foo"),
+///     // TuplePath([1])
+///     (
+///         // TuplePath([1, 0])
+///         (),
+///         // TuplePath([1, 1])
+///         Transform::default(),
+///         // TuplePath([1, 2])
+///         Visibility::Hidden,
+///     ),
+///     // TuplePath([2])
+///     (),
+/// )
+/// ```
+#[derive(Clone, Debug)]
+#[repr(transparent)]
+struct TuplePath(Vec<usize>);
```

Comment:
Beyond implementing `Debug` for this type, nothing else has changed since `insert_unit_bundle` :)